File: sdk/java/src/main/java/io/velocity/examples/CronSchedule.java
package io.velocity.examples;

import io.velocity.sdk.*;

/**
 * Example: Scheduled (cron) workflow using the VELOCITY-WorkFlow Java SDK.
 *
 * Demonstrates:
 *   - Starting a workflow tied to a cron expression
 *   - Simulating a cron fire signal
 *   - Completing the scheduled execution
 *
 * Prerequisites:
 *   1. Start the VELOCITY-WorkFlow server:
 *      cd VELOCITY-WorkFlow/src/Velocity.Workflow.Server && dotnet run
 *   2. Build the SDK:
 *      cd VELOCITY-WorkFlow/sdk/java && ./gradlew build
 *   3. Run this example:
 *      ./gradlew run -PmainClass=io.velocity.examples.CronSchedule
 */
public class CronSchedule {

    private static final String CRON_EXPRESSION = "*/5 * * * *"; // Every 5 minutes

    public static void main(String[] args) {
        System.out.println("=== VELOCITY-WorkFlow Java SDK — Cron Schedule ===\n");

        try (VelocityClient client = VelocityClient.create("localhost:50051")) {
            // 1. Start a workflow with a cron schedule
            WorkflowHandle handle = client.startWorkflow(
                StartWorkflowOptions.builder()
                    .workflowType("periodic-report")
                    .namespace("default")
                    .taskQueue("reports")
                    .totalSteps(1)
                    .build()
            );
            System.out.println("1. Scheduled workflow started: key=" + handle.workflowKey());
            System.out.println("   Cron expression: " + CRON_EXPRESSION);

            // 2. Describe the workflow
            WorkflowDescription desc = client.describeWorkflow(handle.workflowKey());
            System.out.println("2. Status: " + desc.status());

            // 3. Send a cron-fire signal
            boolean signaled = client.signalWorkflow(
                handle.workflowKey(),
                "cron-fire",
                "{\"fire_number\": 1}".getBytes()
            );
            System.out.println("3. Cron fire signal sent: " + signaled);

            // 4. Complete the scheduled execution
            boolean completed = client.completeWorkflow(
                handle.workflowKey(),
                "{\"report\": \"generated\"}".getBytes()
            );
            System.out.println("4. Execution completed: " + completed);
        }

        System.out.println("\n=== Cron schedule example finished! ===");
    }
}

File: sdk/go/examples/cron_schedule.go
// Example: Scheduled (cron) workflow using the VELOCITY-WorkFlow Go SDK.
//
// Demonstrates:
//   - Starting a workflow tied to a cron expression
//   - Simulating a cron fire signal
//   - Completing the scheduled execution
//
// Prerequisites:
//   1. Start the VELOCITY-WorkFlow server:
//      cd VELOCITY-WorkFlow/src/Velocity.Workflow.Server && dotnet run
//   2. go run examples/cron_schedule.go
package main

import (
	"context"
	"fmt"
	"log"
	"time"

	velocity_sdk "github.com/velocity-workflow/sdk/go/velocity_sdk"
)

const cronExpression = "*/5 * * * *" // Every 5 minutes

func main() {
	fmt.Println("=== VELOCITY-WorkFlow Go SDK — Cron Schedule ===")
	fmt.Println()

	ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
	defer cancel()

	client, err := velocity_sdk.NewClient("localhost:50051", "")
	if err != nil {
		log.Fatalf("Failed to connect: %v", err)
	}
	defer client.Close()

	// 1. Start a workflow with a cron schedule
	handle, err := client.StartWorkflow(ctx, &velocity_sdk.StartWorkflowOptions{
		WorkflowType: "periodic-report",
		Namespace:    "default",
		TaskQueue:    "reports",
		TotalSteps:   1,
		Input:        []byte(fmt.Sprintf(`{"cron": "%s"}`, cronExpression)),
	})
	if err != nil {
		log.Fatalf("StartWorkflow failed: %v", err)
	}
	fmt.Printf("1. Scheduled workflow started: key=%d\n", handle.WorkflowKey)
	fmt.Printf("   Cron expression: %s\n", cronExpression)

	// 2. Send a cron-fire signal
	fmt.Println("2. Sending cron-fire signal...")
	// client.SignalWorkflow(ctx, handle.WorkflowKey, "cron-fire", []byte(`{"fire_number": 1}`))

	// 3. Complete the scheduled execution
	fmt.Println("3. Completing scheduled execution...")
	// client.CompleteWorkflow(ctx, handle.WorkflowKey, []byte(`{"report": "generated"}`))

	fmt.Println()
	fmt.Println("=== Cron schedule example finished! ===")
}
